File: app/src/main/java/com/roy/ui/FLens.java
package com.roy.ui;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.SeekBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatSeekBar;
import androidx.fragment.app.Fragment;

import com.roy.R;
import com.roy.itf.LensInterface;
import com.roy.util.UtilSettings;
import com.roy.views.LensView;

public class FLens extends Fragment implements LensInterface {

    LensView lensViewsSettings;
    AppCompatSeekBar sbMinIconSize;
    TextView tvValueMinIconSize;
    AppCompatSeekBar sbDistortionFactor;
    TextView tvValueDistortionFactor;
    AppCompatSeekBar sbScaleFactor;
    TextView tvValueScaleFactor;
    AppCompatSeekBar sbAnimationTime;
    TextView tvValueAnimationTime;

    private UtilSettings utilSettings;

    public FLens() {
    }

    public static FLens newInstance() {
        return new FLens();
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.f_lens, container, false);
        utilSettings = new UtilSettings(getActivity());
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        setupViews(view);
        assignValues();
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        if (getActivity() != null && getActivity() instanceof ASettings) {
            ((ASettings) getActivity()).setLensInterface(this);
        }
    }

    private void setupViews(View view) {
        lensViewsSettings = view.findViewById(R.id.lensViewsSettings);
        sbMinIconSize = view.findViewById(R.id.sbMinIconSize);
        tvValueMinIconSize = view.findViewById(R.id.tvValueMinIconSize);
        sbDistortionFactor = view.findViewById(R.id.sbDistortionFactor);
        tvValueDistortionFactor = view.findViewById(R.id.tvValueDistortionFactor);
        sbScaleFactor = view.findViewById(R.id.sbScaleFactor);
        tvValueScaleFactor = view.findViewById(R.id.tvValueScaleFactor);
        sbAnimationTime = view.findViewById(R.id.sbAnimationTime);
        tvValueAnimationTime = view.findViewById(R.id.tvValueAnimationTime);

        view.findViewById(R.id.sbMinIconSizeParent).setOnClickListener(v -> {
            //do nothing
        });
        view.findViewById(R.id.rlSbDistortionFactorParent).setOnClickListener(v -> {
            //do nothing
        });
        view.findViewById(R.id.rlSbScaleFactorParent).setOnClickListener(v -> {
            //do nothing
        });
        view.findViewById(R.id.rlSbAnimationTimeParent).setOnClickListener(v -> {
            //do nothing
        });

        lensViewsSettings.setDrawType(LensView.DrawType.CIRCLES);
        sbMinIconSize.setMax(UtilSettings.MAX_ICON_SIZE);
        sbMinIconSize.setOnSeekBarChangeListener(new AppCompatSeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                int appropriateProgress = progress + (int) UtilSettings.MIN_ICON_SIZE;
                String minIconSize = appropriateProgress + "dp";
                tvValueMinIconSize.setText(minIconSize);
                utilSettings.save(UtilSettings.KEY_ICON_SIZE, (float) appropriateProgress);
                lensViewsSettings.invalidate();
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
            }
        });
        sbDistortionFactor.setMax(UtilSettings.MAX_DISTORTION_FACTOR);
        sbDistortionFactor.setOnSeekBarChangeListener(new AppCompatSeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                float appropriateProgress = (float) progress / 2.0f + UtilSettings.MIN_DISTORTION_FACTOR;
                String distortionFactor = appropriateProgress + "";
                tvValueDistortionFactor.setText(distortionFactor);
                utilSettings.save(UtilSettings.KEY_DISTORTION_FACTOR, appropriateProgress);
                lensViewsSettings.invalidate();
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
            }
        });
        sbScaleFactor.setMax(UtilSettings.MAX_SCALE_FACTOR);
        sbScaleFactor.setOnSeekBarChangeListener(new AppCompatSeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                float appropriateProgress = (float) progress / 5.0f + UtilSettings.MIN_SCALE_FACTOR;
                String scaleFactor = appropriateProgress + "";
                tvValueScaleFactor.setText(scaleFactor);
                utilSettings.save(UtilSettings.KEY_SCALE_FACTOR, appropriateProgress);
                lensViewsSettings.invalidate();
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
            }
        });
        sbAnimationTime.setMax(UtilSettings.MAX_ANIMATION_TIME);
        sbAnimationTime.setOnSeekBarChangeListener(new AppCompatSeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                long appropriateProgress = (long) progress / 2 + UtilSettings.MIN_ANIMATION_TIME;
                String animationTime = appropriateProgress + "ms";
                tvValueAnimationTime.setText(animationTime);
                utilSettings.save(UtilSettings.KEY_ANIMATION_TIME, appropriateProgress);
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
            }
        });
    }

    private void assignValues() {
        sbMinIconSize.setProgress((int) utilSettings.getFloat(UtilSettings.KEY_ICON_SIZE) - (int) UtilSettings.MIN_ICON_SIZE);
        String minIconSize = (int) utilSettings.getFloat(UtilSettings.KEY_ICON_SIZE) + "dp";
        tvValueMinIconSize.setText(minIconSize);
        sbDistortionFactor.setProgress((int) (2.0f * (utilSettings.getFloat(UtilSettings.KEY_DISTORTION_FACTOR) - UtilSettings.MIN_DISTORTION_FACTOR)));
        String distortionFactor = utilSettings.getFloat(UtilSettings.KEY_DISTORTION_FACTOR) + "";
        tvValueDistortionFactor.setText(distortionFactor);
        sbScaleFactor.setProgress((int) (5.0f * (utilSettings.getFloat(UtilSettings.KEY_SCALE_FACTOR) - UtilSettings.MIN_SCALE_FACTOR)));
        String scaleFactor = utilSettings.getFloat(UtilSettings.KEY_SCALE_FACTOR) + "";
        tvValueScaleFactor.setText(scaleFactor);
        sbAnimationTime.setProgress((int) (2 * (utilSettings.getLong(UtilSettings.KEY_ANIMATION_TIME) - UtilSettings.MIN_ANIMATION_TIME)));
        String animationTime = utilSettings.getLong(UtilSettings.KEY_ANIMATION_TIME) + "ms";
        tvValueAnimationTime.setText(animationTime);
    }

    @Override
    public void onDefaultsReset() {
        resetToDefault();
        assignValues();
    }

    private void resetToDefault() {
        utilSettings.save(UtilSettings.KEY_ICON_SIZE, UtilSettings.DEFAULT_ICON_SIZE);
        utilSettings.save(UtilSettings.KEY_DISTORTION_FACTOR, UtilSettings.DEFAULT_DISTORTION_FACTOR);
        utilSettings.save(UtilSettings.KEY_SCALE_FACTOR, UtilSettings.DEFAULT_SCALE_FACTOR);
        utilSettings.save(UtilSettings.KEY_ANIMATION_TIME, UtilSettings.DEFAULT_ANIMATION_TIME);
    }
}

File: app/src/main/java/com/mckimquyen/ui/FLens.kt
package com.mckimquyen.ui

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import android.widget.TextView
import androidx.appcompat.widget.AppCompatSeekBar
import androidx.fragment.app.Fragment
import com.mckimquyen.R
import com.mckimquyen.enums.DrawType
import com.mckimquyen.itf.LensInterface
import com.mckimquyen.util.UtilSettings
import com.mckimquyen.views.LensView

class FLens : Fragment(), LensInterface {
    companion object {
        fun newInstance(): FLens {
            return FLens()
        }
    }

    @JvmField
    var lensViewsSettings: LensView? = null

    @JvmField
    var sbMinIconSize: AppCompatSeekBar? = null

    @JvmField
    var tvValueMinIconSize: TextView? = null

    @JvmField
    var sbDistortionFactor: AppCompatSeekBar? = null

    @JvmField
    var tvValueDistortionFactor: TextView? = null

    @JvmField
    var sbScaleFactor: AppCompatSeekBar? = null

    @JvmField
    var tvValueScaleFactor: TextView? = null

    @JvmField
    var sbAnimationTime: AppCompatSeekBar? = null

    @JvmField
    var tvValueAnimationTime: TextView? = null
    private var utilSettings: UtilSettings? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val view = inflater.inflate(R.layout.frm_lens, container, false)
        utilSettings = UtilSettings(requireContext())
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        setupViews(view)
        assignValues()
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (activity != null && activity is ASettings) {
            (activity as ASettings?)?.setLensInterface(this)
        }
    }

    private fun setupViews(view: View) {
        lensViewsSettings = view.findViewById(R.id.lensViewsSettings)
        sbMinIconSize = view.findViewById(R.id.sbMinIconSize)
        tvValueMinIconSize = view.findViewById(R.id.tvValueMinIconSize)
        sbDistortionFactor = view.findViewById(R.id.sbDistortionFactor)
        tvValueDistortionFactor = view.findViewById(R.id.tvValueDistortionFactor)
        sbScaleFactor = view.findViewById(R.id.sbScaleFactor)
        tvValueScaleFactor = view.findViewById(R.id.tvValueScaleFactor)
        sbAnimationTime = view.findViewById(R.id.sbAnimationTime)
        tvValueAnimationTime = view.findViewById(R.id.tvValueAnimationTime)

        view.findViewById<View>(R.id.sbMinIconSizeParent).setOnClickListener { }
        view.findViewById<View>(R.id.rlSbDistortionFactorParent).setOnClickListener { }
        view.findViewById<View>(R.id.rlSbScaleFactorParent).setOnClickListener { }
        view.findViewById<View>(R.id.rlSbAnimationTimeParent).setOnClickListener { }

        lensViewsSettings?.setDrawType(DrawType.CIRCLES)

        sbMinIconSize?.let { sb ->
            sb.max = UtilSettings.MAX_ICON_SIZE
            sb.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
                override fun onProgressChanged(
                    seekBar: SeekBar,
                    progress: Int,
                    fromUser: Boolean
                ) {
                    val appropriateProgress = progress + UtilSettings.MIN_ICON_SIZE.toInt()
                    val minIconSize = appropriateProgress.toString() + "dp"
                    tvValueMinIconSize?.text = minIconSize
                    utilSettings?.save(UtilSettings.KEY_ICON_SIZE, appropriateProgress.toFloat())
                    lensViewsSettings?.invalidate()
                }

                override fun onStartTrackingTouch(seekBar: SeekBar) {}
                override fun onStopTrackingTouch(seekBar: SeekBar) {}
            })
        }

        sbDistortionFactor?.let { sb ->
            sb.max = UtilSettings.MAX_DISTORTION_FACTOR
            sb.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
                override fun onProgressChanged(
                    seekBar: SeekBar,
                    progress: Int,
                    fromUser: Boolean
                ) {
                    val appropriateProgress =
                        progress.toFloat() / 2.0f + UtilSettings.MIN_DISTORTION_FACTOR
                    val distortionFactor = appropriateProgress.toString() + ""
                    tvValueDistortionFactor?.text = distortionFactor
                    utilSettings?.save(UtilSettings.KEY_DISTORTION_FACTOR, appropriateProgress)
                    lensViewsSettings?.invalidate()
                }

                override fun onStartTrackingTouch(seekBar: SeekBar) {}
                override fun onStopTrackingTouch(seekBar: SeekBar) {}
            })
        }

        sbScaleFactor?.let { sb ->
            sb.max = UtilSettings.MAX_SCALE_FACTOR
            sb.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
                override fun onProgressChanged(
                    seekBar: SeekBar,
                    progress: Int,
                    fromUser: Boolean
                ) {
                    val appropriateProgress =
                        progress.toFloat() / 5.0f + UtilSettings.MIN_SCALE_FACTOR
                    val scaleFactor = appropriateProgress.toString() + ""
                    tvValueScaleFactor?.text = scaleFactor
                    utilSettings?.save(UtilSettings.KEY_SCALE_FACTOR, appropriateProgress)
                    lensViewsSettings?.invalidate()
                }

                override fun onStartTrackingTouch(seekBar: SeekBar) {}
                override fun onStopTrackingTouch(seekBar: SeekBar) {}
            })
        }

        sbAnimationTime?.let { sb ->
            sb.max = UtilSettings.MAX_ANIMATION_TIME
            sb.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
                override fun onProgressChanged(
                    seekBar: SeekBar,
                    progress: Int,
                    fromUser: Boolean
                ) {
                    val appropriateProgress =
                        progress.toLong() / 2 + UtilSettings.MIN_ANIMATION_TIME
                    val animationTime = appropriateProgress.toString() + "ms"
                    tvValueAnimationTime?.text = animationTime
                    utilSettings?.save(UtilSettings.KEY_ANIMATION_TIME, appropriateProgress)
                }

                override fun onStartTrackingTouch(seekBar: SeekBar) {}
                override fun onStopTrackingTouch(seekBar: SeekBar) {}
            })
        }
    }

    private fun assignValues() {
        utilSettings?.let { us ->
            sbMinIconSize?.progress =
                us.getFloat(UtilSettings.KEY_ICON_SIZE).toInt() - UtilSettings.MIN_ICON_SIZE.toInt()
            val minIconSize =
                us.getFloat(UtilSettings.KEY_ICON_SIZE).toInt().toString() + "dp"
            tvValueMinIconSize?.text = minIconSize
            sbDistortionFactor?.progress =
                (2.0f * (us.getFloat(UtilSettings.KEY_DISTORTION_FACTOR) - UtilSettings.MIN_DISTORTION_FACTOR)).toInt()

            val distortionFactor =
                us.getFloat(UtilSettings.KEY_DISTORTION_FACTOR).toString() + ""
            tvValueDistortionFactor?.text = distortionFactor
            sbScaleFactor?.progress =
                (5.0f * (us.getFloat(UtilSettings.KEY_SCALE_FACTOR) - UtilSettings.MIN_SCALE_FACTOR)).toInt()

            val scaleFactor = us.getFloat(UtilSettings.KEY_SCALE_FACTOR).toString() + ""
            tvValueScaleFactor?.text = scaleFactor
            sbAnimationTime?.progress =
                (2 * (us.getLong(UtilSettings.KEY_ANIMATION_TIME) - UtilSettings.MIN_ANIMATION_TIME)).toInt()
            val animationTime =
                us.getLong(UtilSettings.KEY_ANIMATION_TIME).toString() + "ms"
            tvValueAnimationTime?.text = animationTime
        }
    }

    override fun onDefaultsReset() {
        resetToDefault()
        assignValues()
    }

    private fun resetToDefault() {
        utilSettings?.let { us ->
            us.save(
                /* name = */ UtilSettings.KEY_ICON_SIZE,
                /* value = */ UtilSettings.DEFAULT_ICON_SIZE
            )
            us.save(
                /* name = */ UtilSettings.KEY_DISTORTION_FACTOR,
                /* value = */ UtilSettings.DEFAULT_DISTORTION_FACTOR
            )
            us.save(
                /* name = */ UtilSettings.KEY_SCALE_FACTOR,
                /* value = */ UtilSettings.DEFAULT_SCALE_FACTOR
            )
            us.save(
                /* name = */ UtilSettings.KEY_ANIMATION_TIME,
                /* value = */ UtilSettings.DEFAULT_ANIMATION_TIME
            )
        }
    }
}
